### PR TITLE
cxx-qt-gen: initial preparation for closures in signals

### PR DESCRIPTION
**Pull request by ahayzen-kdab**
Related to #595 

This is the first commits from #681

### CODE REVIEW

File: crates/cxx-qt-gen/src/parser/mod.rs [hunk 1]
```diff
@@ -252,4 +252,41 @@ mod tests {
         let parser = Parser::from(module);
         assert!(parser.is_err());
     }
+
+    #[test]
+    fn test_cxx_qobject_namespace() {
+        let module: ItemMod = parse_quote! {
+            #[cxx_qt::bridge(namespace = "bridge_namespace")]
+            mod ffi {
+                extern "RustQt" {
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/parser/mod.rs:261]: done as an additional commit

File: crates/cxx-qt-gen/src/writer/cpp/header.rs [hunk 2]
```diff
@@ -130,7 +132,10 @@ pub fn write_cpp_header(generated: &GeneratedCppBlocks) -> String {
     .fold(generated.includes.clone(), |mut acc, qobject| {
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/writer/cpp/header.rs:132]: I find this whole iterator chain a bit confusing.

If I understand it correctly, the point is to create a BTreeSet of includes and then convert that to a list of strings.
However, the way it's written what's happening is this:
Iterate over the list of QObjects, create a BTreeSet of includes out of this.
Then discard that BTreeSet by converting it to an iterator again, then chain that iterator, fold it again into a new BTreeSet, just to then collect it again.

Could you maybe rewrite it to something like this (pseudocode):
```rust
let mut include_set = BTreeSet::new();
include_set.extend(
    generated.qobjects.iter()
         .flat_map(|qobject|qobject.blocks.includes)
         .cloned()
         .chain(genereated.extern_cxx_qt.iter().flat_map(block::includes).cloned())
);
include_set.into_iter().collect::<Vec<String>>().join("\n");
```
That would move the semantics to:
Create a BTreeSet out of this list of includes.
Then convert it to a `Vec<String>`.

Which at least to me seems a lot clearer. Maybe that's just me though.

File: crates/cxx-qt-gen/include/cxxqt_signalhandler.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#pragma once
+
+#include <type_traits>
+
+#include "rust/cxx.h"
+
+namespace rust::cxxqtlib1 {
+
+template<typename CXXArguments>
+class SignalHandler
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/include/cxxqt_signalhandler.h:17]: Maybe add a comment here that this represents a Rust `Box<dyn FnMut>` by deferring all actual operations to Rust, apart from moving it, which is done in terms of raw memory copy, which is equivalent to what Rust does.

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]
I think we can hide some of our implementation detail in our public documentation.

Apart from that mostly nitpicks. Overall direction is sane :)

File: crates/cxx-qt/src/cxxqtsignalhandler.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    /// A mutable reference to the inner closure
+    pub fn closure(&mut self) -> &mut Box<T::FnType> {
+        &mut self.closure
+    }
+}
+
+// Safety:
+//
+// Static checks on the C++ and Rust side to ensure the size is the same.
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/cxxqtsignalhandler.rs:45]: Look at the next change :-) we have static_assertions in Rust :tada:

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/cxxqtsignalhandler.rs:45]: We end up generating

```rust
cxx_qt::assert_eq_align!(
    cxx_qt::CxxQtSignalHandler<MyObjectCxxQtSignalClosureready>,
    usize
);
cxx_qt::assert_eq_size!(
    cxx_qt::CxxQtSignalHandler<MyObjectCxxQtSignalClosureready>,
    [usize; 2]
);
```

File: crates/cxx-qt/Cargo.toml [hunk 5]
```diff
@@ -20,6 +20,7 @@ categories = ["api-bindings", "gui"]
 [dependencies]
 cxx.workspace = true
 cxx-qt-macro.workspace = true
+static_assertions = "1.1.0"
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/Cargo.toml:23]: It's only re-exported at this point for use by cxx-qt-gen later

File: crates/cxx-qt/src/lib.rs [hunk 6]
```diff
@@ -9,13 +9,20 @@
 //!
 //! See the [book](https://kdab.github.io/cxx-qt/book/) for more information.
 
+mod cxxqtsignalhandler;
 mod cxxqtthread;
 
 pub use cxx_qt_macro::bridge;
 pub use cxx_qt_macro::qobject;
 
+pub use cxxqtsignalhandler::{CxxQtSignalHandler, CxxQtSignalHandlerClosure};
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs:18]: So then should the whole signalhandler module be public instead ? So you do `cxx_qt::signalhandler::QtSignalHandler}` ?

File: crates/cxx-qt/src/lib.rs [hunk 7]
```diff
@@ -9,13 +9,20 @@
 //!
 //! See the [book](https://kdab.github.io/cxx-qt/book/) for more information.
 
+mod cxxqtsignalhandler;
 mod cxxqtthread;
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs:13]: Right these are always namespaced to `cxx_qt::T` as they are in the crate so it's fine to be simplified. I think the names came from where we do `type T` in CXX bridges and we try to prefix them, but not necessary here

File: crates/cxx-qt/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs:18]: done

File: crates/cxx-qt/src/lib.rs [hunk 8]
```diff
@@ -9,12 +9,19 @@
...
 pub use cxx_qt_macro::bridge;
 pub use cxx_qt_macro::qobject;
 
-pub use cxxqtthread::CxxQtThread;
+pub use threading::CxxQtThread;
+
+// Export static assertions that can then be used in cxx-qt-gen generation
+//
+// These are currently used to ensure that CxxQtSignalHandler has the right size
+pub use static_assertions::{assert_eq_align, assert_eq_size};
```
[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply crates/cxx-qt/src/lib.rs:24]: Side note for the future:
We should also try to re-export `cxx`.
As I've made the mistake a couple of times to not include `cxx` as a dependency in the `Cargo.toml`, which leads to very weird errors.

We'd probably have to always generate `use cxx_qt::cxx` in every `#[cxx_qt::bridge]` such that the generated `#[cxx::bridge]` works, which isn't great, but still better than having this somewhat obfuscated dependency everywhere...

Something for later though.

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs]: yeah good idea, maybe write an issue so we don't forget

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs]: changed to re-export all of `static_assertions`

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

**Pull request merged**